circuit-maintenance UIComponent

Pavan-183 opened the pull request:
<!--
    Thank you for your interest in contributing to Circuit Maintenance! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->

# Closes: #<ISSUE NUMBER GOES HERE>

## What's Changed

<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->
| View                                        | Before                                                                                                                                                                                   | After                                                                                                                                                                                    |
| ------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **List View (Circuit Maintenances)**        | ![image](https://github.com/user-attachments/assets/c57bd94e-3b4e-4aac-8668-61006f981654) <br> ![image](https://github.com/user-attachments/assets/1b69af3d-560f-4904-9153-d634a5b49280) | <img width="1919" height="669" alt="image" src="https://github.com/user-attachments/assets/a18d25f3-97af-44fe-8e2e-349742831726" /> <br> <img width="1707" height="288" alt="image" src="https://github.com/user-attachments/assets/0d5919cd-9341-4b72-b12c-25d1d0e8c9ae" /> |
| **Detail View (Circuit Maintenances)**      | ![image](https://github.com/user-attachments/assets/c9d6be8e-7a98-4121-b8ea-0ae78524cc10)                                                                                                | <img width="1735" height="866" alt="image" src="https://github.com/user-attachments/assets/86585d9a-c206-4035-8a43-99df0e1f7849" /> <br> <img width="1713" height="606" alt="image" src="https://github.com/user-attachments/assets/12240a18-837f-422e-949b-7b0f8ce73d3a" /> <br><img width="1712" height="365" alt="image" src="https://github.com/user-attachments/assets/103c6f10-0718-400e-9641-828fcc9207bc" />
| **Update View (Circuit Maintenances)**      | ![image](https://github.com/user-attachments/assets/f1af096e-039d-47ee-b7d4-e272ff9fa4c4) <br> ![image](https://github.com/user-attachments/assets/9979b550-924a-46c0-b89a-8f87240f70da) | <img width="1585" height="956" alt="image" src="https://github.com/user-attachments/assets/b680805f-e281-4d51-9b90-8281733c6e3f" /> <br> <img width="1722" height="820" alt="image" src="https://github.com/user-attachments/assets/6bff08bc-7388-4df4-8f8a-3f575aa29e71" /> |
| **Delete View (Circuit Maintenances)**      | ![image](https://github.com/user-attachments/assets/b3786450-18d4-4bc9-b016-d7891312cd9e) <br> ![image](https://github.com/user-attachments/assets/2b52638e-188b-4c8f-892e-024c46f09645) | <img width="1497" height="418" alt="image" src="https://github.com/user-attachments/assets/d303caff-6558-4915-b661-b6797f478af8" /> <br> <img width="1734" height="625" alt="image" src="https://github.com/user-attachments/assets/b821260a-7049-4316-925f-256c032db210" /> |
| **Bulk Edit View (Circuit Maintenances)**   | ![image](https://github.com/user-attachments/assets/1c2ccb83-a9c8-429b-a53c-08e20c97cccf) <br> ![image](https://github.com/user-attachments/assets/d369185c-0ef0-47a7-bd2a-859009a57d35) | <img width="1723" height="832" alt="image" src="https://github.com/user-attachments/assets/ac0296b8-a230-4baa-a1fc-1d77994f2010" /> <br> <img width="1726" height="902" alt="image" src="https://github.com/user-attachments/assets/6f284eea-103f-4f7d-ab3d-e87b5dc47702" /> |
| **Bulk Delete View (Circuit Maintenances)** | ![image](https://github.com/user-attachments/assets/ec77c4e1-12ad-4902-b11b-735bac6c06da) <br> ![image](https://github.com/user-attachments/assets/1d8f3532-6647-4fba-9b0c-ffabaaf96f11) | <img width="1554" height="476" alt="image" src="https://github.com/user-attachments/assets/4f0e7989-7c6c-4341-aad4-df18dc131eb0" /> <br> <img width="1727" height="939" alt="image" src="https://github.com/user-attachments/assets/0a09ada0-c69b-4491-aae9-0411a1eb0c95" /> |
| **Add View (Circuit Maintenances)**         | ![image](https://github.com/user-attachments/assets/29541669-aeff-4f66-bfb2-f04c47d33c0d) <br> ![image](https://github.com/user-attachments/assets/527a64b2-a941-4f00-955b-e8f28d1ec117) | <img width="1728" height="966" alt="image" src="https://github.com/user-attachments/assets/1180cc03-c499-4cea-9ec9-0613d18690a7" /> <br> <img width="1770" height="673" alt="image" src="https://github.com/user-attachments/assets/ad437e52-a5f7-457c-b9e6-c305c3f2bb63" /> |
| **Add View (Circuit Impact)**               | ![image](https://github.com/user-attachments/assets/d1df9bcf-34e5-4a21-82dd-6c545ac536be) <br> ![image](https://github.com/user-attachments/assets/e690e390-d87c-406f-9ccd-b2bd5e1c4554) | <img width="1610" height="724" alt="image" src="https://github.com/user-attachments/assets/ee55e600-9053-4830-8f6b-7134c01a24b7" /> <br> <img width="1769" height="797" alt="image" src="https://github.com/user-attachments/assets/e9252734-8c94-4e91-9046-32583f9328ce" /> |
| **Add View (Note)**                         | ![image](https://github.com/user-attachments/assets/8a623919-999d-47ad-b448-6aa9f4df4571) <br> ![image](https://github.com/user-attachments/assets/787ddfa3-57a5-4afa-ade8-33a45a1cfaca) | <img width="1714" height="898" alt="image" src="https://github.com/user-attachments/assets/d54a432c-0b17-4df2-8554-456ec813cea8" /> <br> <img width="1770" height="970" alt="image" src="https://github.com/user-attachments/assets/5ad58e3e-efc0-44ca-a044-da754c96c9b8" /> |
| **List View (Notifications)**               | ![image](https://github.com/user-attachments/assets/7a5ce3f2-aa38-467f-9a21-09654ae76e9f) <br> ![image](https://github.com/user-attachments/assets/92328480-0668-4037-aec8-3402b58af0b4) | <img width="1731" height="581" alt="image" src="https://github.com/user-attachments/assets/9dd46055-58d6-4a2e-ba77-12c12082c5d3" /> <br> <img width="1716" height="250" alt="image" src="https://github.com/user-attachments/assets/5134df75-a7b8-4416-9345-f10b9d6cf0e5" /> |
| **Detail View (Notifications)**             | ![image](https://github.com/user-attachments/assets/a01de4d7-2c85-4152-95c0-6870983a03ac)                                                                                                | <img width="1727" height="829" alt="image" src="https://github.com/user-attachments/assets/19a53b26-8368-4545-9cdf-7d0dca71e645" /> |
| **List View (Notification Source)**         | ![image](https://github.com/user-attachments/assets/d01d5072-9cd4-4bdd-b45f-d1f81fae886d)                                                                                                | <img width="1726" height="719" alt="image" src="https://github.com/user-attachments/assets/6cae7e65-3a69-49c6-aa89-e0b45b32d9ca" /> |
| **Detail View (Notification Source)**       | ![image](https://github.com/user-attachments/assets/c767d175-4390-4748-92da-625bacde2a3d)                                                                                                | <img width="1735" height="708" alt="image" src="https://github.com/user-attachments/assets/db1e058a-a97c-48f6-aca0-2a2b91880902" /> |
| **Update View (Notification Source)**       | NA                                                                                                                                                                                       | <img width="1583" height="548" alt="image" src="https://github.com/user-attachments/assets/c98a4120-c99f-4d6b-ae0c-3ac3adb803fe" /> <br> <img width="1738" height="855" alt="image" src="https://github.com/user-attachments/assets/d778a2bd-3c62-4d80-8ab9-67ed8c39252f" /> |
| **Bulk Update View (Notification Source)**  | ![image](https://github.com/user-attachments/assets/8d42ce5a-ddca-4c4a-b163-20f21fcb8717)                                                                                                | <img width="1727" height="758" alt="image" src="https://github.com/user-attachments/assets/cbf1c6b4-f434-4828-9458-c1bdc0996820" /> <br> <img width="1756" height="966" alt="image" src="https://github.com/user-attachments/assets/fc6c982c-4681-45c1-9653-aa96c47ce9c3" /> |
| **Bulk Delete View (Notification Source)**  | NA                                                                                                                                                                                       | <img width="1160" height="404" alt="image" src="https://github.com/user-attachments/assets/c9992728-7011-485f-83e7-aa6e2e4cb35a" /> <br> <img width="1729" height="920" alt="image" src="https://github.com/user-attachments/assets/dcd6e0b6-5b76-4499-a2d5-b9575963df68" /> |
| **Delete View (Notification Source)**       | NA                                                                                                                                                                                       | <img width="1103" height="312" alt="image" src="https://github.com/user-attachments/assets/c8debc25-b8a1-40fe-bbe2-22491a1a93bd" /> |

## To Do

<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

Pavan-183 committed:
circuit-maintenance UIComponent

Pavan-183 committed:
Fixed pylint issue

Sridevi-NTC commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuitimpact_retrieve.html:
Do we need this file?

Pavan-183 replied:
Removed @Sridevi-NTC

Sridevi-NTC commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuitimpact.html:
```diff
@@ -1,27 +1,2 @@
-    </div>
-{% endblock %}
+{% extends 'nautobot_circuit_maintenance/circuitimpact_retrieve.html' %}
+{% comment %}3.0 TODO: remove this template, which only exists for backward compatibility with 2.4 and earlier{% endcomment %}
```
Can we use {% extends 'generic/object_detail.html' %} instead?

Sridevi-NTC commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/rawnotification_retrieve.html:
Is this template required?

Pavan-183 replied:
Removed @Sridevi-NTC

Sridevi-NTC commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuitmaintenance_retrieve.html:
```diff
@@ -0,0 +1,7 @@
+{% extends 'generic/object_detail.html' %}
+{% load helpers %}
+
+{% block masthead %}
+    <h1>Maintenance ID: {% block title %}{{ object }}{% endblock %}</h1>
+{% endblock masthead %}
```
@itdependsnetworks Do you think we actually need the {% block masthead %} section?

jvanderaa replied:
What is block masthead?

Pavan-183 replied:
The {% block masthead %} section defines the header area of the maintenance detail page.
Due to this block, we’re able to display the “Maintenance ID: ” 
heading at the top of the page (as shown in the UI).

<img width="931" height="553" alt="image" src="https://github.com/user-attachments/assets/2c1318de-d5fc-4edb-a283-9cd592e2af52" />

Sridevi-NTC commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/notificationsource.html:
```diff
@@ -1,62 +1,2 @@
-        {% endif %}
-    </div>
-{% endblock %}
+{% extends 'nautobot_circuit_maintenance/notificationsource_retrieve.html' %}
```
Can't we extend object_detail?

Sridevi-NTC commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/notificationsource_retrieve.html:
Do we need this template?

Sridevi-NTC commented on nautobot_circuit_maintenance/tables.py:
```diff
@@ -59,15 +75,44 @@ class CircuitImpactTable(BaseTable):
-    impact = tables.Column(linkify=True)
+    # maintenance = tables.Column(linkify=True)
+    # circuit = tables.Column(linkify=True)
+    # impact = tables.Column(linkify=True)
```
Is there a specific reason for this change?

Pavan-183 replied:
Previously, with the old fields, the table only showed Maintenance, Circuit, and Impact columns. Because of that, important details such as Provider, Type, Status, A Side, Z Side, and Description were missing so for testing purpose I have commented out those fields now I have uncommented these.

Sridevi-NTC commented on nautobot_circuit_maintenance/tables.py:
```diff
@@ -81,11 +126,25 @@ class Meta(BaseTable.Meta):
 
 class NotificationSourceTable(BaseTable):
-    """Table to display Circuit Impact model."""
+    """Table to display CircuitImpact model."""
```
Notifications?

Sridevi-NTC commented:
@Pavan-183 Could you add the data for notifications and update the screenshot?

Sridevi-NTC commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -189,80 +209,84 @@ def get_maintenances_per_month(self):
+                table_title="Circuits",
+                exclude_columns=["actions", "dynamic_groups"],
+                add_button_route="plugins:nautobot_circuit_maintenance:circuitimpact_add",
+                paginate=None,
+                show_table_config_button=None,
```
Don't we need pagination and config button?

Pavan-183 replied:
<img width="1683" height="531" alt="image" src="https://github.com/user-attachments/assets/2de84a9a-36c4-42ad-9014-e6f742af5d6f" />

If I don't include those, it looks like this @Sridevi-NTC , So I'm not sure whether to keep them or not.

Pavan-183 replied:
I have addressed [this here](https://github.com/nautobot/nautobot-app-circuit-maintenance/pull/347#discussion_r2447509306).

Sridevi-NTC commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -189,80 +209,84 @@ def get_maintenances_per_month(self):
+                related_field_name="maintenance",
+                table_title="Circuits",
+                exclude_columns=["actions", "dynamic_groups"],
+                add_button_route="plugins:nautobot_circuit_maintenance:circuitimpact_add",
```
Is Add button not created automatically?

Pavan-183 replied:
Yes, the Add button is created automatically. I’ve removed the explicit add_button_route.

Sridevi-NTC commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -189,80 +209,84 @@ def get_maintenances_per_month(self):
+                related_field_name="maintenance",
+                exclude_columns=["actions", "dynamic_groups"],
+                add_button_route="plugins:nautobot_circuit_maintenance:note_add",
+                paginate=None,
+                show_table_config_button=None,
```
Don't we need pagination and config button? Is Add button not created automatically?

Pavan-183 replied:
Without `paginate` and `show_table_config_button`, the tables appear like this (screenshot attached).
I wasn’t sure if we want to keep the pagination and configure button visible in this section, so I’ve currently set both parameters to None. Please confirm if we should display them as in the default view.

<img width="1674" height="550" alt="Screenshot 2025-10-13 142240" src="https://github.com/user-attachments/assets/944e0f4e-c7ba-4b39-9715-bbeb267747aa" />

Sridevi-NTC commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -189,80 +209,84 @@ def get_maintenances_per_month(self):
-        maintenance_note = models.Note.objects.filter(maintenance=instance)
-        circuits = models.CircuitImpact.objects.filter(maintenance=instance)
-        parsednotification = models.ParsedNotification.objects.filter(maintenance=instance).order_by(
-            "-raw_notification__stamp"
```
Missing order_by("-raw_notification__stamp") in UI component

Sridevi-NTC commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -273,241 +297,220 @@ def get(self, request, *args, **kwargs):
+        if key != "raw":
+            return super().render_value(key, value, context)
+        # Normalize to bytes
+        if isinstance(value, memoryview):
```
Any particular reason for checking memoryview rather than bytes?

Pavan-183 replied:
I think some binary fields come as `memoryview` objects, so this check converts them to bytes for consistency. I'm not completely sure though, I’ll do a bit more investigation here.

Sridevi-NTC commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -273,241 +297,220 @@ def get(self, request, *args, **kwargs):
-                raw_repr = instance.raw.tobytes().decode("utf-8", "strict")
-        except UnicodeDecodeError as exc:
-            raw_repr = "Raw content was not able to be decoded with utf-8"
-            logger.warning("%s: %s", raw_repr, exc)
```
Shouldn’t this be logged as a warning in the UI component?

Pavan-183 commented:
> @Pavan-183 Could you add the data for notifications and update the screenshot?

sure @Sridevi-NTC

Pavan-183 committed:
Addressed few comments

jvanderaa commented on .github/workflows/ci.yml:
```diff
@@ -105,8 +105,8 @@ jobs:
       matrix:
-        python-version: ["3.11"]
-        nautobot-version: ["2.0.0"]
+        python-version: ["3.12"]
+        nautobot-version: ["2.4.16"]
```
This change should probably be made as a separate PR.

jvanderaa commented on .github/workflows/ci.yml:
```diff
@@ -151,13 +151,13 @@ jobs:
       fail-fast: true
       matrix:
-        python-version: ["3.8"]  # 3.12 stable is tested in unittest_report stage.
+        python-version: ["3.9"]  # 3.12 stable is tested in unittest_report stage.
```
Why change this on UIComponent?

jvanderaa commented on development/nautobot_config.py:
```diff
@@ -143,7 +143,7 @@
                 "account": os.environ.get("CM_NS_1_ACCOUNT", ""),
                 "secret": os.environ.get("CM_NS_1_SECRET", ""),
-                "url": os.environ.get("CM_NS_1_URL", ""),
+                "url": os.environ.get("CM_NS_1_URL", "https"),
```
This doesn't make sense.

jvanderaa commented on nautobot_circuit_maintenance/tables.py:
```diff
@@ -1,9 +1,25 @@
+from nautobot.core.tables import (
+    BaseTable,
+    ToggleColumn,
+)
```
Undo, no functional change.

jvanderaa commented on nautobot_circuit_maintenance/tables.py:
```diff
@@ -63,11 +79,39 @@ class CircuitImpactTable(BaseTable):
 
         model = CircuitImpact
-        fields = ("pk", "maintenance", "circuit", "impact")  # pylint:disable=nb-use-fields-all
+        fields = (
+            "cid",
+            "impact",
+            "provider",
+            "circuit_type",
+            "status",
+            "circuit_termination_a",
+            "circuit_termination_z",
+            "description",
+        )
+        default_columns = fields
```
What is with this change compared to the previous?

Pavan-183 replied:
Previously, with the old fields, the table only showed Maintenance, Circuit, and Impact (as shown in the below image). Because of that, important details such as Provider, Type, Status, A Side, Z Side, and Description were missing.

<img width="1708" height="189" alt="image" src="https://github.com/user-attachments/assets/a8093ef1-699f-42eb-b706-b2bfb30ba534" />

With this update, the table now correctly displays all relevant fields (as shown in the below image), which provides a more complete view of the circuit impact.

<img width="1690" height="190" alt="image" src="https://github.com/user-attachments/assets/00ec7bf8-e252-40e0-a6c7-a7a9d518c2bf" />

jvanderaa commented on pyproject.toml:
```diff
@@ -44,7 +44,7 @@ pydantic = ">=1.10.4,<3"
 [tool.poetry.group.dev.dependencies]
 coverage = "*"
-django-debug-toolbar = "*"
+django-debug-toolbar = "5.2.0"
```
No, this should not be changed in this fashion.

jvanderaa reviewed (commented):
Please make sure that there are a few updates made, such as the django-debug-toolbar. I will not set to request changes with being out of office for a while. cc: @itdependsnetworks

Pavan-183 committed:
Addressed comments

Pavan-183 commented:
<img width="1322" height="925" alt="image" src="https://github.com/user-attachments/assets/e366026a-ea9c-48a7-beb0-051f9c4c884d" />

Hi @Sridevi-NTC ,

I’ve addressed all the comments, but I think I accidentally started a review instead of replying individually. I’ll go back and respond to each comment properly.
Sorry for the inconvenience, and thank you for your patience!

Pavan-183 commented:
> @Pavan-183 Could you add the data for notifications and update the screenshot?

Added @Sridevi-NTC , I manually created demo Raw Notification and Maintenance because Parsed Notification requires parent objects. Since no real emails were parsed, I inserted a fake Parsed Notification using test JSON to show how it looks in the UI.

Pavan-183 committed:
Addressed PR comments

jvanderaa commented on pyproject.toml:
```diff
@@ -44,7 +44,7 @@ pydantic = ">=1.10.4,<3"
 [tool.poetry.group.dev.dependencies]
 coverage = "*"
-django-debug-toolbar = "*"
+django-debug-toolbar = "5.2.0"
```
Please revert.

Pavan-183 replied:
@jvanderaa , can you [review this one](https://github.com/nautobot/nautobot-app-circuit-maintenance/pull/350/) please

glennmatthews commented:
Is this replaced by #350, or how does it relate?

Pavan-183 commented:
> Is this replaced by #350, or how does it relate?

@glennmatthews , the new PR #350 replaces the old one, based on Ken's suggestion, I pulled the squashed changes and opened a new PR.
